File: app/src/main/java/com/spiderbiggen/randomchampionselector/storage/database/daos/ChampionDAO.java
package com.spiderbiggen.randomchampionselector.storage.database.daos;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.spiderbiggen.randomchampionselector.model.Champion;

import java.util.Collection;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Single;

/**
 * On 26-2-2018.
 *
 * @author Stefan Breetveld
 */
@Dao
public interface ChampionDAO {

    @Query("SELECT * FROM champion WHERE roles LIKE :role||'%' ORDER BY name")
    Flowable<List<Champion>> getAll(String role);

    @Query("SELECT * FROM champion ORDER BY name")
    Flowable<List<Champion>> getAll();

    @Query("SELECT DISTINCT roles FROM champion")
    Flowable<List<String>> getAllRoles();

    @Query("SELECT * FROM champion WHERE roles LIKE :role||'%' ORDER BY RANDOM() LIMIT 1")
    Flowable<Champion> getRandom(String role);

    @Query("SELECT * FROM champion ORDER BY RANDOM() LIMIT 1")
    Flowable<Champion> getRandom();

    @Query("SELECT * FROM champion WHERE `key` = :id")
    Flowable<Champion> getChampion(int id);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insert(Collection<Champion> entities);

    @Update(onConflict = OnConflictStrategy.IGNORE)
    void update(Collection<Champion> entities);

    default void insertAll(Collection<Champion> entities) {
        insert(entities);
        update(entities);
    }

    @Delete
    void delete(Champion champion);

}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/storage/database/daos/ChampionDAO.kt
package com.spiderbiggen.randomchampionselector.storage.database.daos

import android.arch.persistence.room.*

import com.spiderbiggen.randomchampionselector.model.Champion

import io.reactivex.Flowable

/**
 * On 26-2-2018.
 *
 * @author Stefan Breetveld
 */
@Dao
interface ChampionDAO {

    @get:Query("SELECT * FROM champion ORDER BY name")
    val all: Flowable<List<Champion>>

    @get:Query("SELECT DISTINCT roles FROM champion")
    val allRoles: Flowable<List<String>>

    @get:Query("SELECT * FROM champion ORDER BY RANDOM() LIMIT 1")
    val random: Flowable<Champion>

    @Query("SELECT * FROM champion WHERE roles LIKE '%'||:role||'%' ORDER BY name")
    fun getAll(role: String?): Flowable<List<Champion>>

    @Query("SELECT * FROM champion WHERE roles LIKE '%'||:role||'%' ORDER BY RANDOM() LIMIT 1")
    fun getRandom(role: String?): Flowable<Champion>

    @Query("SELECT * FROM champion WHERE `key` = :id")
    fun getChampion(id: Int): Flowable<Champion>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(entities: Collection<Champion>)

    @Update(onConflict = OnConflictStrategy.IGNORE)
    fun update(entities: Collection<Champion>)

    @Delete
    fun delete(champion: Champion)

    @Delete
    fun deleteAll(champion: Collection<Champion>)

}
